src/main/java/com/mongodb/spark/sql/connector/connection/MongoClientCache.java

```diff
@@ -0,0 +1,179 @@
+    scheduler.scheduleWithFixedDelay(
+        this::checkClientCache, initialCleanUpDelayMS, cleanUpDelayMS, TimeUnit.MILLISECONDS);
+    isAvailable = true;
+  }
+
+  /**
+   * Acquires a new MongoClient
+   *
+   * @param mongoClientFactory the factory to use for creating the MongoClient
+   * @return the MongoClient from the cache or create a new one using the {@code
+   *     MongoClientFactory}.
+   */
+  synchronized MongoClient acquire(final MongoClientFactory mongoClientFactory) {
+    assertIsAvailable();
+    return cacheList.stream()
+        .filter(c -> Objects.equals(c.mongoClientFactory, mongoClientFactory))
+        .findFirst()
+        .orElseGet(
+            () -> {
+              MongoClientCacheItem mongoClientCacheItem =
+                  new MongoClientCacheItem(mongoClientFactory, keepAliveNanos);
+              cacheList.add(mongoClientCacheItem);
+              return mongoClientCacheItem;
+            })
+        .acquire();
+  }
+
+  /**
+   * Releases the MongoClient
+   *
+   * @param mongoClient to release
+   */
+  synchronized void release(final MongoClient mongoClient) {
+    assertIsAvailable();
+    cacheList.stream()
+        .filter(c -> Objects.equals(c.mongoClient, mongoClient))
+        .findFirst()
+        .orElseThrow(() -> new IllegalStateException("MongoClient not in the cache."))
+        .release();
+  }
+
+  void shutdown() {
+    if (isAvailable) {
+      scheduler.shutdown();
+      cacheList.forEach(
+          item -> {
+            try {
+              item.mongoClient.close();
+            } catch (Exception e) {
+              LOGGER.info("Error when shutting down client: {}", e.getMessage());
+            }
+          });
+      cacheList.clear();
+      isAvailable = false;
+    }
+  }
+
+  private synchronized void checkClientCache() {
+    long now = System.nanoTime();
+    cacheList.removeIf(cacheItem -> cacheItem.shouldBeRemoved(now));
+  }
+
+  private void assertIsAvailable() {
+    Assertions.ensureState(
+        () -> isAvailable, "The MongoClientCache has been shutdown and is no longer available");
+  }
+
+  private static final class MongoClientCacheItem {
+    private final MongoClientFactory mongoClientFactory;
+    private final long keepAliveNanos;
+    private final MongoClient mongoClient;
+    private long releaseTime;
+    private int referenceCount;
+
+    private MongoClientCacheItem(
+        final MongoClientFactory mongoClientFactory, final long keepAliveNanos) {
+      this.mongoClientFactory = mongoClientFactory;
+      this.mongoClient = mongoClientFactory.create();
+      this.keepAliveNanos = keepAliveNanos;
+      this.releaseTime = Long.MAX_VALUE;
+      this.referenceCount = 0;
+    }
+
+    private MongoClient acquire() {
+      referenceCount += 1;
+      releaseTime = Long.MAX_VALUE;
+      return mongoClient;
+    }
+
+    private void release() {
+      referenceCount -= 1;
```

Comment:
Since nothing stops the code that uses `MongoClientCache` from releasing more than once per acquire, let's have an assertion here that ensures `referenceCount` does not go below 0.

Reply:
Done.